kwok/cloudprovider/helpers.go

```diff
@@ -30,78 +32,102 @@ import (
+func ConstructInstanceTypes() ([]*cloudprovider.InstanceType, error) {
+	var instanceTypes []*cloudprovider.InstanceType
+	var instanceTypeOptions []InstanceTypeOptions
+
+	if err := json.Unmarshal(rawInstanceTypes, &instanceTypeOptions); err != nil {
+		return nil, fmt.Errorf("could not parse JSON data: %w", err)
 	}
-	return map[string]string{
-		InstanceSizeLabelKey:   size,
-		InstanceFamilyLabelKey: family,
-		InstanceCPULabelKey:    fmt.Sprintf("%d", cpu),
-		InstanceMemoryLabelKey: fmt.Sprintf("%d", cpu*memFactor*1024),
+
+	for _, opts := range instanceTypeOptions {
+		opts = setDefaultOptions(opts)
+		instanceTypes = append(instanceTypes, newInstanceType(opts))
 	}
+	return instanceTypes, nil
 }
 
-// InstanceTypesAssorted create many unique instance types with varying CPU/memory/architecture/OS/zone/capacity type.
-func ConstructInstanceTypes() []*cloudprovider.InstanceType {
-	var instanceTypes []*cloudprovider.InstanceType
-	for _, cpu := range []int{1, 2, 4, 8, 16, 32, 48, 64, 96, 128, 192, 256} {
-		for _, memFactor := range []int{2, 4, 8} {
-			for _, os := range []sets.Set[string]{sets.New(string(v1.Linux)), sets.New(string(v1.Windows))} {
-				for _, arch := range []string{v1beta1.ArchitectureAmd64, v1beta1.ArchitectureArm64} {
-					// Construct instance type details, then construct offerings.
-					mem := cpu * memFactor
-					pods := lo.Clamp(cpu*16, 0, 1024)
-					labels := MakeInstanceTypeLabels(cpu, memFactor)
-					opts := InstanceTypeOptions{
-						Name:             fmt.Sprintf("%s-%s-%s-%s", labels[InstanceFamilyLabelKey], labels[InstanceSizeLabelKey], arch, strings.Join(sets.List(os), ",")),
-						Architecture:     arch,
-						OperatingSystems: os,
-						Resources: v1.ResourceList{
-							v1.ResourceCPU:              resource.MustParse(fmt.Sprintf("%d", cpu)),
-							v1.ResourceMemory:           resource.MustParse(fmt.Sprintf("%dGi", mem)),
-							v1.ResourcePods:             resource.MustParse(fmt.Sprintf("%d", pods)),
-							v1.ResourceEphemeralStorage: resource.MustParse("20G"),
-						},
-						InstanceTypeLabels: labels,
-					}
-					price := PriceFromResources(opts.Resources)
-
-					opts.Offerings = cloudprovider.Offerings{}
-					for _, zone := range KwokZones {
-						for _, ct := range []string{v1beta1.CapacityTypeSpot, v1beta1.CapacityTypeOnDemand} {
-							opts.Offerings = append(opts.Offerings, cloudprovider.Offering{
-								CapacityType: ct,
-								Zone:         zone,
-								Price:        lo.Ternary(ct == v1beta1.CapacityTypeSpot, price*.7, price),
-								Available:    true,
-							})
-						}
-					}
-					instanceTypes = append(instanceTypes, newInstanceType(opts))
-				}
-			}
+// parseSizeFromType will attempt to discover the instance size if it matches a special AWS format.
+// If not, fall back to the cpu value.  This works for both Azure and GCP (and the "generic" instances
+// generated by tools/gen_instances.go)
+func parseSizeFromType(ty, cpu string) string {
```

Comment:
musing: What if you could just add these labels yourself? Rather than having to parse out the data from the instance type naming, it seems more flexible for anyone trying to integrate their own special form of instance types into this kwok provider if you just opened up a "labels" or "requirements" block for instance types and just allowed everyone to set their own custom "labels" for nodes that were launched or selected on with that instance type.

This seems good enough to me for now, but might be a cool follow-on if we could make this a bit more flexible.

Reply:
@jonathan-innis I thought about this and almost added the functionality, but "my" interpretation was that if users wanted to add their own labels that should be done as part of the NodePool config?